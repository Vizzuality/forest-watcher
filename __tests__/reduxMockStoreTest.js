import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

global.fetch = require('jest-fetch-mock');

const mockStore = configureMockStore([thunk]);

// TEST

describe('async actions', () => {
  afterEach(() => {
    fetch.resetMocks();
  });

  it('creates FETCH_TODOS_SUCCESS when fetching todos has been done', () => {
    const initialState = { todos: ['ONE'] };
    const store = mockStore(initialState);
    const responseBody = { todos: ['TWO', 'THREE'] };

    fetch.mockResponseOnce(JSON.stringify(responseBody));

    const expectedActions = [{ type: FETCH_TODOS_REQUEST }, { type: FETCH_TODOS_SUCCESS, body: responseBody }];

    return store.dispatch(fetchTodos()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_TODOS_FAILURE when fetching todos is rejected', () => {
    const store = mockStore();
    const errorObject = { name: 'Mock Error', message: 'Mock error message.' };

    fetch.mockRejectOnce(errorObject);

    const expectedActions = [{ type: FETCH_TODOS_REQUEST }, { type: FETCH_TODOS_FAILURE, exception: errorObject }];

    return store.dispatch(fetchTodos()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});

// REDUX ACTIONS

const FETCH_TODOS_REQUEST = 'FETCH_TODOS_REQUEST';
const FETCH_TODOS_SUCCESS = 'FETCH_TODOS_SUCCESS';
const FETCH_TODOS_FAILURE = 'FETCH_TODOS_FAILURE';

function fetchTodosRequest() {
  return {
    type: FETCH_TODOS_REQUEST
  };
}

function fetchTodosSuccess(body) {
  return {
    type: FETCH_TODOS_SUCCESS,
    body
  };
}

function fetchTodosFailure(exception) {
  return {
    type: FETCH_TODOS_FAILURE,
    exception
  };
}

export function fetchTodos() {
  return dispatch => {
    dispatch(fetchTodosRequest());
    return fetch('http://example.com/todos')
      .then(res => res.json())
      .then(body => dispatch(fetchTodosSuccess(body)))
      .catch(exception => dispatch(fetchTodosFailure(exception)));
  };
}
