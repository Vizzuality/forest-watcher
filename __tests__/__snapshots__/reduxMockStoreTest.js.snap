// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates FETCH_TODOS_FAILURE when fetching todos is rejected 1`] = `
Array [
  Object {
    "type": "FETCH_TODOS_REQUEST",
  },
  Object {
    "exception": Object {
      "message": "Mock error message.",
      "name": "Mock Error",
    },
    "type": "FETCH_TODOS_FAILURE",
  },
]
`;

exports[`async actions creates FETCH_TODOS_SUCCESS when fetching todos has been done 1`] = `
Array [
  Object {
    "type": "FETCH_TODOS_REQUEST",
  },
  Object {
    "body": Object {
      "todos": Array [
        "TWO",
        "THREE",
      ],
    },
    "type": "FETCH_TODOS_SUCCESS",
  },
]
`;
